### PR TITLE
more detailed Applicability results / add metric to ConstraintResult / scoverage

### PR DESCRIPTION
**Pull request by tdhd**
*Description of changes:*
* each `ConstraintResult` in `CheckResult` is now linked to the metric that the constraint evaluation was based on
* the result of `Applicability.isApplicable(check)` now contains all constraints and their applicabilities
* adds *scoverage* with optional code coverage requirements, currently we are at a method coverage of 81% (run with `mvn scoverage:report`, https://github.com/scoverage/scoverage-maven-plugin)
* test for constraint assertions that throw exceptions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala [hunk 1]
```diff
@@ -133,34 +146,38 @@ private[deequ] class Applicability(session: SparkSession) {
...
-    val constraintsByName = check.constraints
+    val namedMetrics = check.constraints
       .map { constraint => constraint.toString -> constraint }
       .map {
         case (name, nc: ConstraintDecorator) => name -> nc.inner
         case (name, c: Constraint) => name -> c
       }
       .collect { case (name, constraint: AnalysisBasedConstraint[_, _, _]) =>
-        (name, constraint)
+        (name, constraint.analyzer.calculate(data).value)
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala]: maybe refactor this into a temporary variable

File: src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala [hunk 2]
```diff
@@ -133,34 +146,38 @@ private[deequ] class Applicability(session: SparkSession) {
...
       }
 
+    val constraintApplicabilities = check.constraints.zip(namedMetrics).map {
+      case (constraint, (_, metric)) =>
+        constraint -> metric.isSuccess
+    }.toMap
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala]: .toMap should be on the next line

File: src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala [hunk 3]
```diff
@@ -170,7 +187,7 @@ private[deequ] class Applicability(session: SparkSession) {
     * @param schema The schema of the data the analyzers are for
     */
   def isApplicable(analyzers: Seq[Analyzer[_ <: State[_], Metric[_]]], schema: StructType)
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala]: params should be on their own lines

File: src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala [hunk 4]
```diff
@@ -55,46 +55,45 @@ private[deequ] case class AnalysisBasedConstraint[S <: State[S], M, V](
...
+        try {
+          val assertOn = runPickerOnMetric(metricValue)
+          val assertionOk = runAssertion(assertOn)
+
+          if (assertionOk) {
+            ConstraintResult(this, ConstraintStatus.Success, metric = Some(metric))
+          } else {
+            var errorMessage = s"Value: $assertOn does not meet the constraint requirement!"
+
+            hint.foreach(hint => errorMessage += s" $hint")
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala]: use mkString

File: src/main/scala/com/amazon/deequ/constraints/Constraint.scala [hunk 5]
```diff
@@ -49,8 +50,8 @@ class ConstraintDecorator(protected val _inner: Constraint) extends Constraint {
       analysisResults: Map[Analyzer[_, Metric[_]], Metric[_]])
     : ConstraintResult = {
 
-    val result = _inner.evaluate(analysisResults)
-    ConstraintResult(this, result.status, result.message)
+    // constraint = this to keep toString of NamedConstraint
```
Comment (sscdotopen) [inline src/main/scala/com/amazon/deequ/constraints/Constraint.scala]: ?

File: src/test/scala/com/amazon/deequ/VerificationSuiteTest.scala [hunk 6]
```diff
@@ -57,24 +56,28 @@ class VerificationSuiteTest extends WordSpec with Matchers with SparkContextSpec
           CheckStatus.Success)
         assert(VerificationSuite().onData(df).addCheck(checkToErrorOut).run().status ==
```
Comment (sscdotopen) [inline src/test/scala/com/amazon/deequ/VerificationSuiteTest.scala]: refactor VerificationSuite().onData(df).addCheck(checkToErrorOut).run().status into a temp variable  (also for the cases below)

File: src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala (hunk 2 above)
Comment (tdhd) [reply src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala]: Still haven't lost hope for myself that I'll learn deequ style :)

Do you think we could have a rule for it in our style check?

File: src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala (hunk 4 above)
Comment (tdhd) [reply src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala]: :+1:

File: src/main/scala/com/amazon/deequ/constraints/Constraint.scala (hunk 5 above)
Comment (tdhd) [reply src/main/scala/com/amazon/deequ/constraints/Constraint.scala]: Should have put a better description there. Most of the constraints are `NamedConstraint`s. `_inner` in this case is an `AnalysisBasedConstraint` which doesn't have a pretty-printable name. To keep the name and toString method, we keep the outer type.

Will update the comment.

File: src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala (hunk 4 above)
Comment (tdhd) [reply src/main/scala/com/amazon/deequ/constraints/AnalysisBasedConstraint.scala]: First thought mkString would make it more readable and compact. However also `None.mkString(" ", "", "")` prints a leading space. I'll keep the foreach instead.

File: src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala (hunk 2 above)
Comment (sscdotopen) [reply src/main/scala/com/amazon/deequ/analyzers/applicability/Applicability.scala]: We tried to configure such rules, unfortunately, there is always an edge case that is not coverable by the stylechecker...

**Comment by sscdotopen** [comment]
@stefan-grafberger Thank you for having a look the changes!

**Pull request merged**